core/src/main/java/fr/sncf/osrd/new_infra_state/api/NewTrainPath.java

```diff
@@ -45,13 +48,41 @@ public String toString() {
     /** Utility function, returns the last element with offset <= the given offset.
      * For range objects (route / sections), it returns the element containing the offset. */
     public static <T> T getLastElementBefore(ImmutableList<LocatedElement<T>> list, double offset) {
+        return getLastLocatedElementBefore(list, offset).element;
+    }
+
+    /** Utility function, returns the last located element with offset <= the given offset.
+     * For range objects (route / sections), it returns the element containing the offset. */
+    public static <T> LocatedElement<T> getLastLocatedElementBefore(
+            ImmutableList<LocatedElement<T>> list,
+            double offset
+    ) {
         for (int i = 0; i < list.size(); i++) {
             var e = list.get(i);
             if (e.pathOffset > offset && i > 0)
-                return list.get(i - 1).element;
+                return list.get(i - 1);
             if (e.pathOffset >= offset)
-                return e.element;
+                return e;
+        }
+        return list.get(list.size() - 1);
+    }
+
+    /** Converts a track location into an offset on the path.
+     * Note: using this should be avoided whenever possible, it prevents paths with loops */
+    public double convertTrackLocation(TrackLocation location) {
+        for (var track : trackRangePath)
+            if (track.element.contains(location))
+                return track.pathOffset + track.element.offsetOf(location);
+        throw new InvalidSchedule("TrackLocation isn't included in the path");
+    }
+
+    /** Returns the location at the given offset */
+    public TrackLocation findLocation(double pathOffset) {
+        var element = getLastLocatedElementBefore(trackRangePath, pathOffset);
+        if (element.element.track.getEdge() instanceof SwitchBranch) {
+            // This case can happen when pathOffset is exactly on a switch, we want the next range in that case
+            element = trackRangePath.get(trackRangePath.indexOf(element) + 1);
         }
-        return list.get(list.size() - 1).element;
+        return element.element.locationOfOffset(pathOffset - element.pathOffset);
```

Comment:
It's not *that* simple, because we can have objects with 0 length.

It would be nice to have such an object but it would be a fairly large change, how about we move this to a new issue?

Reply:
agreed